File: apps/app_weatherman/weatherman_0.3/src/weatherman/WeatherMan.java
package weatherman;
import java.util.Timer;
import java.io.IOException;

import io.APIKeyReader;

/**
 * WeatherMan gives live weather information about a location, 
 * gathered from the WeatherUnderground API. This program makes one API
 * call every 20 minutes. 
 * 
 * @author Mike
 * @version 0.3.0
 *
 */
public class WeatherMan {
	
	// TODO: Failed connection calls should try again, fast.
	
//	******************************************************
//					TOP LEVEL DJINNI
//	******************************************************
	
	/**
	 * The version number of this application.
	 */
	public static final String VERSION = "0.3.1";
	
	/**
	 * The root folder of the Djinni Home AI System's program data
	 */
	public static final String ROOT_FILE_PATH = "C:\\ProgramData\\Djinni\\";

	
	
//	******************************************************
//					WEATHERMAN APP
//	******************************************************
	
	
	/** 
	 * The file name of the Weather Underground API key 
	 */
	public static final String WEATHER_UNDERGROUND_KEY_NAME = "wu_key.txt";
	
	/**
	 * The file name the WeatherReports will be saved under
	 */
	public static final String WEATHER_REPORT_FILE_NAME = "wu_report.txt";
	
	/** 
	 * Key to the WeatherUnderground API. It is read from a text file.
	 * @see APIKeyReader
	 */
	private static String api_key;
	
	public static void main(String[] args) throws InterruptedException{
		System.out.println("*** WEATHERMAN APP " + VERSION + " ***");
		System.out.println("");
		
		try{
			// Get the API key from the Djinni Program Data folder
			api_key = new APIKeyReader( WEATHER_UNDERGROUND_KEY_NAME).getKey();
			
			// Set up a timer that will gather Weather Data on a fixed interval
			Timer time = new Timer();
			WeatherTask wt = new WeatherTask(api_key);
			time.schedule(wt, 0, WeatherTask.INTERVAL);
			
		} catch( IOException e) {
			// If the API key can not be loaded, log it in the console. 
			// Note that the timer is not started, and URL connection is never attempted.
			System.out.println("Error: WeatherMan could not locate your API Key.");
			System.out.println("The API Key should be stored in " + ROOT_FILE_PATH + WEATHER_UNDERGROUND_KEY_NAME);
		}
	}
}

File: apps/weatherman/weatherman_0.4/Weatherman/ConsoleApplication1/WeatherMan.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

using Module.Vox;

namespace Djinni.Apps.WeatherManApp
{
    class WeatherMan
    {
        #region Vars & Consts 
        /// <summary>
        /// The version of this app.
        /// </summary>
        public const String VERSION = "0.4.1";    

        /// <summary>
        ///  Key to the WU API.
        /// </summary>
        private String api_key;

        #endregion

        #region Constructor
        /// <summary>
        /// Create a Weatherman app. This app is capable of gathering weather data, and speaking it to the user.
        /// </summary>
        /// <param name="vox">Module in charge of voice out.</param>
        /// <param name="API_key_filepath">The folder which contains the WeatherUnderground API key.</param>
        /// <param name="API_key_filename">The name of the non-encrypted text file which contains the WeatherUnderground API key.</param>
        public WeatherMan(String API_key_filepath, String API_key_filename)
        {
            System.Console.WriteLine("Loading App: WeatherMan v{0}", VERSION);
            System.Console.WriteLine("    Reading API Key...");

            try
            {
                // Get the API key from the Djinni Program Data folder
                api_key = new APIKeyReader(API_key_filepath, API_key_filename).getKey();
                System.Console.WriteLine("    API key found.");
                System.Console.WriteLine("    App loaded successfully.");
            }
            catch (Exception e)
            {
                // If the API key cannot be read, log the error in the console and throws an error.
                // Note, that a URL connection is never attempted.
                String errorMessage = "WeatherMan could not locate your API key.";
                Console.WriteLine("    Error: {0}", errorMessage);
                Console.WriteLine("    {0}", e.Message);
                System.Console.WriteLine("    App could not be loaded.");
                throw new IOException(errorMessage);
            }
        }

        /// <summary>
        /// Create a Weatherman app. This app is capable of gathering weather data, and speaking it to the user.
        /// </summary>
        /// <param name="vox">Module in charge of voice out.</param>
        /// <param name="API_key">String of the WeatherUnderground API key.</param>
        public WeatherMan(String API_key)
        {
            System.Console.WriteLine("Loading App: WeatherMan v{0}", VERSION);
            api_key = API_key;
            System.Console.WriteLine("    App loaded successfully.");
        }
        #endregion

        #region Public Methods
        /// <summary>
        /// Creates a WeatherReport in the console, by pinging the WeatherUnderground API.
        /// <para/>This method makes 2 calls to the WU API. The API can handle up to 10 calls/min, and 500 calls/day.
        /// </summary>
        public void CreateWeatherReport()
        {
            WeatherReport myReport = new WeatherReport(api_key);
            myReport.print();
        }

        /// <summary>
        /// Creates a WeatherReport, spoken out loud, and logged in the console, by pinging the WeatherUnderground API.
        /// <para/>This method makes 2 calls to the WU API. The API can handle up to 10 calls/min, and 500 calls/day.
        /// </summary>
        /// <param name="vox">A Vox module, that will speak the report.</param>
        public void CreateWeatherReport(Vox vox)
        {
            WeatherReport myReport = new WeatherReport(api_key);
            
            // print the report into the console.
            myReport.print();

            // speak the line over the Vox.
            vox.Queue(myReport.ToString());
        }
        #endregion
    }
}
